### ISSUE TITLE
Structured clone data when using `pushState`

### ISSUE THREAD

**Issue by lucasfcosta** [jsdom/jsdom#2970]
Hi all, before getting to the issue, just wanna say thanks for your amazing work.

### Summary

Currently, when using `pushState`, the `history` will hold a reference to the pushed object. It should, instead, do a structured clone.

Holding references can cause `history.state` have the wrong content if the previously pushed object has suffered any mutation.

### Basic info:

- **Node.js version:** v10.16.0 (not important for this specific bug)
- **jsdom version:** v16.2.2

### Minimal reproduction case

```js
const { JSDOM } = require("jsdom");
const dom = new JSDOM(`<!DOCTYPE html><p>History Example</p>`);
const history = dom.window.history;

const state = { name: "Bruce" };

history.pushState(state, dom.window.document.title);

state.name = { name: "Batman" };

history.pushState(state, dom.window.document.title);


dom.window.addEventListener("popstate", () => {
  // Logs "Batman", not "Bruce"
  console.log(history.state.name);
});

history.back();
```

### How does similar code behave in browsers?

Running the exact same code on a browser (without the Node.js specifics), causes "Bruce" to be logged:

```js
const state1 = { name: "Bruce" };

history.pushState(state, window.document.title);

state.name = { name: "Batman" };

history.pushState(state, window.document.title);

window.addEventListener("popstate", () => {
  // Logs "Bruce" on browsers
  console.log(history.state.name);
});

history.back();
```

### Extra details

I've seen there's a `TODO` [here](https://github.com/jsdom/jsdom/blob/2b0dbc98bcffa738c676dc99d49d5716e5578422/lib/jsdom/living/window/History-impl.js#L69), indicating that a _structured clone_ should happen, which would fix this problem.

A structured clone before the `data` assignment [here](https://github.com/jsdom/jsdom/blob/2b0dbc98bcffa738c676dc99d49d5716e5578422/lib/jsdom/living/window/History-impl.js#L110) would solve this.

I'm currently writing a book which relies on JSDOM for a similar example, and I'd be willing to submit a fix for this very soon if you agree, as I depend on that.

I would just like to know if you've got any specific requirements/concerns in regards to the structured cloning algorithm to use.

Best,
Lucas

**Comment by domenic** [comment]
Thanks for the nice bug report!

> I would just like to know if you've got any specific requirements/concerns in regards to the structured cloning algorithm to use.

The requirements are basically the usual ones. Follow the spec, and find or write web platform tests.

The spec in question here is https://html.spec.whatwg.org/multipage/structured-data.html#safe-passing-of-structured-data (most notably for this particular example https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeforstorage and its dependencies).

A good amount of tests are available in https://github.com/web-platform-tests/wpt/tree/master/html/webappapis/structured-clone and https://github.com/web-platform-tests/wpt/tree/master/html/infrastructure/safe-passing-of-structured-data, although to pass those we'd probably need to add structured cloning to postMessage to. The filenames in https://github.com/web-platform-tests/wpt/tree/master/html/browsers/history/the-history-interface are uninformative, but hopefully one of them contains tests specifically in the context of `history.pushState`...

**Comment by bsides** [comment]
TL;DR: Is this bug related to the lack of testable state after changing a route? 

Explaning: right now I'm facing this problem

```javascript
history.push('/somewhere', {state1})

// ...then somewhere along the same page, after an user interaction

history.push('/somewhere', {state2})
```

I'm running jest + testing library + react and it never gets the data from state2.
